inst/include/population_dynamics/population/population.hpp

```diff
@@ -33,15 +33,457 @@
+              (1 - exp(-(this->mortality_Z[index_ya])));
+            this -> catch_numbers_at_age[index_yaf] += catch_;
+            fleets[fleet_] -> catch_numbers_at_age[index_yaf] += catch_;
+          }
+        }
+
+        /**
+         * @brief Calculates expected catch weight at age for each fleet for a given year and age
+         * 
+         * @param year the year of expected catch weight at age
+         * @param age the age of expected catch weight at age
+         */
+        void CalculateCatchWeightAA(int year, int age) {
+          for (size_t fleet_=0; fleet_ < this -> nfleets; fleet_++) {
+            int index_yaf = year * this -> nages * this -> nfleets + age * this -> nfleets + fleet_;
+            this->catch_weight_at_age[index_yaf] =
+              this->catch_numbers_at_age[index_yaf] *
+              this -> weight_at_age[age];
+          }
+        }
+
+        /**
+         * @brief Calculates expected proportion of individuals mature at a selected ageage
+         * 
+         * @param index_ya dimension folded index for year and age
+         * @param age the age of maturity
+         */
+        void CalculateMaturityAA(int index_ya, int age) {
+          //this->maturity is pointing to the maturity module, which has
+          // an evaluate function. -> can be nested.
+          this -> proportion_mature_at_age[index_ya] = 
+            this -> maturity -> evaluate(age);
+        }
+
+        /**
+         * @brief Executes the population loop
+         * 
+         */
+        void Evaluate() { 
+        
+        /*
+          Sets derived vectors to zero
+          Performs parameters transformations 
+          Sets recruitment deviations to mean 0, and then adds bias adjustment.
+        */
+         Prepare();      
+         /*
+          start at year=0, age=0; 
+          here year 0 is the estimated initial stock structure and age 0 are recruits
+          loops start at zero with if statements inside to specify unique code for 
+          initial structure and recruitment 0 loops. Could also have started loops at
+          1 with initial structure and recruitment setup outside the loops.
+
+          year loop is extended to <= nyears because SSB is calculted as the start of the year 
+          value and by extending one extra year we get estimates of the population structure
+          at the end of the final year. An alternative approach would be to keep initial numbers
+          at age in it's own vector and each year to include the population structure at the end of
+          the year. This is likely a null point given that we are planning to modify to an event/stanza
+          based structure in later milestones which will elimitate this confusion by explicity
+          referencing the exact date (or period of averaging) at which any calculation or output is 
+          being made. 
+         */
+         for (size_t y = 0; y <= this->nyears; y++) {
```

Comment:
I saw heap corruption errors (Exit code 0xc0000374) when testing the CalculateMortality() and CalculateNumbersAA functions using year <= nyears in for loops. The error disappeared if using year<nyears. Details can be found [here](https://github.com/NOAA-FIMS/FIMS/blob/013abfd36e713cc8a35fb7fb3df635603d3e508d/tests/gtest/test_population_Li.cpp#L10-L19). Please let me know if the error is related to the discussion here and if you have a suggestion on if year < nyears or year <= nyears should be used in the test. Thanks!

Reply:
Hey Bai, The error for CalculateMortality() is because mortality is only nyears long so it should be <nyears. For the numbers at age and SSB tests it should work but we need to get the indexing of these and mortality correct.